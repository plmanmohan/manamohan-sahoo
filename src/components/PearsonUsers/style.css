.user-container {
  background-color: #d8d8d8;
  box-sizing: border-box;
  width: 100%;
  height: 600px;
  padding: 20px 20px;
}
.loader {
  position: "absolute";
  left: "50%";
  top: "50%";
  text-align: center;
  padding: 50px;
  margin: 50px;
}
.heading {
  font-size: 36px;
  color: #007fa3;
  text-shadow: 2px 2px 3px rgba(112, 212, 240, 0.8);
  font-weight: normal;
  padding: 20px 20px;
  margin: 0px;
}
