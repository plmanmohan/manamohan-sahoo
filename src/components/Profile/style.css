.container {
  background-color: #fff;
  box-sizing: border-box;
  width: 212px;
  padding: 10px 10px;
  position: relative;
  box-shadow: 0 0 5px 5px #ccc;
  margin: 50px 18px 20px 19px;
  float: left;
}

.avatar-img {
  width: 80px;
  height: 80px;
  position: absolute;
  top: -40px;
  border-radius: 50%;
  left: 50%;
  margin-left: -40px;
}

.userName {
  font-size: 18px;
  font-weight: normal;
  color: #e93939;
  text-align: center;
  margin: 50px 0 25px 0;
}

.delete {
  background: none;
  border: none;
  font-size: 14px;
  color: #007fa3;
  cursor: pointer;
}

.delete:hover {
  color: #024b5f;
}
