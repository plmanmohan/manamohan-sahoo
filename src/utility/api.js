export default function api() {
    return fetch("https://reqres.in/api/users?page=1&per_page=10", {
             method: "get"
            }).then(response => response.json())
            .catch(error => error);
}