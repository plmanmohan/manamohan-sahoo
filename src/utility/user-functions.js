/**
 * It search for the particular element with the id and removes the element from the Object
 * @param {*Object} state
 * @param {*number} id
 */
export function removeUser(state, id) {
  const remainder = state.users.filter(user => (user.id !== id ? user : null));
  return remainder;
}
/**
 * It merges both of the arrays and removes the duplicate elements
 * @param {*} oldUserList
 * @param {*} newUserList
 */
export function removeDuplicateUsers(oldUserList, newUserList) {
  if (newUserList.length === 0) {
    return oldUserList;
  }
  console.log("State.user :: ", oldUserList);
  var newData = oldUserList.concat(newUserList);
  console.log("Concat :: ", newData);
  
  const uniqueList = [];
  const finalData = [];

  for (let i = 0; i < newData.length; i++) {
    let fullName = newData[i].first_name + newData[i].last_name;
    if (uniqueList.indexOf(fullName) === -1) {
      uniqueList.push(fullName);
      finalData.push(newData[i]);
    }
  }

  console.log("AfterMerge :: ", finalData);
  return finalData;
}
