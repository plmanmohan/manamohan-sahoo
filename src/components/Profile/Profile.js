import React from 'react';

import './style.css';

export const Profile = ({ user, onDelete}) => (
  <div>
    <div className="container">
      <img
        className="avatar-img"
        src={user.avatar}
        alt={user.first_name}
      />
      <p className="userName">
        {user.first_name}
        {' '}
        {user.last_name}
      </p>
      <button type="button" className="delete" onClick={() => onDelete(user.id)}>
        Delete
      </button>
    </div>
  </div>
);
