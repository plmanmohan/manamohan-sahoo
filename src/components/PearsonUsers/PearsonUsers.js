import React, { Component } from 'react';
import { Profile } from '../Profile';
import { removeUser, removeDuplicateUsers } from '../../utility/user-functions';
import './style.css';
import api from '../../utility/api';

export class PearsonUsers extends Component {
  /**
   * Here initial state is defined
   * @param {*}
   */
  constructor() {
    super();
    this.state = {
      users: [
        {
          id: 4,
          first_name: 'Eve',
          last_name: 'Holt',
          avatar:
            'https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg'
        },
        {
          id: 5,
          first_name: 'Charles',
          last_name: 'Morris',
          avatar:
            'https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg'
        },
        {
          id: 6,
          first_name: 'Tracey',
          last_name: 'Ramos',
          avatar:
            'https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg'
        }
      ],
      isLoading: false,
      error: false
    };

    this.removeUser = this.removeUser.bind(this);
  }


  componentDidMount() {
    this.fetchTheData();
  }

  fetchTheData() {
    this.setState({ isLoading: true });
    api()
      .then((result) => {
        const updatedUserList = this.removeDuplicateUsers(result.data);
        this.setState({ users: updatedUserList, isLoading: false });
      })
      .catch(() => {
        this.setState({ error: true, isLoading: false });
      });
  }

  removeDuplicateUsers(userList) {
    return removeDuplicateUsers(this.state.users, userList);
  }

  displayUsers() {
    return this.state.users.map(user => (
      <Profile key={user.id} user={user} onDelete={this.removeUser} />
    ));
  }

  removeUser(id) {
    this.setState({ users: removeUser(this.state, id) });
  }

  render() {
    const { isLoading, error } = this.state;
    return (
      <div className="pearon-users">
        <h1 className="heading">Pearson User Management </h1>
        <div className="user-container">
          {isLoading ? (
            <div className="loader">
              <img
                height={80}
                width={80}
                src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                alt="loading"
              />
            </div>
          ) : (
            this.displayUsers()
          )}
          {error ? 'Error in loading' : ''}
        </div>
      </div>
    );
  }
}
