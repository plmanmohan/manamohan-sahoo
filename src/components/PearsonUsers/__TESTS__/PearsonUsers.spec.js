import { Profile } from "../../Profile";
import { PearsonUsers } from "../PearsonUsers";
import { removeUser, removeDuplicateUsers } from "../../../utility/user-functions";


jest.mock('../../../utility/api.js');

describe('Test PearsonUsers component', () => {

  it('renders without crashing', () => {
    shallow(<PearsonUsers />);
  });

  it("Render <PearsonUsers/>", () => {
    const wrapper = shallow( <PearsonUsers /> );
   expect(wrapper).toMatchSnapshot();
    const h1 = wrapper.find("h1");
    expect(h1.text()).toEqual("Pearson User Management ");
  });
  
  //remove duplicated users from the state add only unique into state
  it("removeDuplicateUsers removes duplicates adds unique users into state and ", () => {
    const oldUserList = [
      { id: 2, first_name: "Janet", last_name: "Weaver", avatar: "128.jpg" },
      { id: 3, first_name: "Emma", last_name: "Wong", avatar: "128.jpg" }
    ];
    const newUserList = [
      { id: 12, first_name: "Rachel", last_name: "Howell", avatar: "128.jpg" },
      { id: 2, first_name: "Janet", last_name: "Weaver", avatar: "128.jpg" }
    ];
    const finalUserList = removeDuplicateUsers(oldUserList, newUserList);
    expect(finalUserList).toEqual([
      { id: 2, first_name: "Janet", last_name: "Weaver", avatar: "128.jpg" },
      { id: 3, first_name: "Emma", last_name: "Wong", avatar: "128.jpg" },
      { id: 12, first_name: "Rachel", last_name: "Howell", avatar: "128.jpg" }
    ]);
  });

  //removeUser
  it("removeUser removes the user", () => {
    const state = {
      users: [                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
        {
          id: 2,
          first_name: "Janet",
          last_name: "Weaver",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg"
        },
        {
          id: 3,
          first_name: "Emma",
          last_name: "Wong",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/olegpogodaev/128.jpg"
        },
        {
          id: 4,
          first_name: "Eve",
          last_name: "Holt",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        }
      ]
    };
    const id = 3;
    const finalState = removeUser(state, id);
    expect(finalState).toEqual([
      {
        id: 2,
        first_name: "Janet",
        last_name: "Weaver",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg"
      },
      {
        id: 4,
        first_name: "Eve",
        last_name: "Holt",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
      }
    ]);
  });

  it("Render <profile/>", () => {
    const state = {
      users: [
        { id: 11, first_name: "George", last_name: "Edwards", avatar: "128.jpg" }
      ]
    };
    const wrapper = shallow(
      <Profile key={11} user={state.users[0]} onDelete={() => {}} />
    );
  
    const p = wrapper.find("p");
    expect(p.text()).toEqual("George Edwards");
  });

  it("Render initial state data", () => {
    const wrapper = shallow(<PearsonUsers />);
    const finalState = wrapper.instance().displayUsers();
    expect(finalState.length).toEqual(3);
  });
 });

