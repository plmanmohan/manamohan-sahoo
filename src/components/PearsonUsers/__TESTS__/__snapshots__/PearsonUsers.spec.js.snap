// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PearsonUsers component Render <PearsonUsers/> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PearsonUsers />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="heading"
>
          Pearson User Management 
</h1>,
        <div
          className="user-container"
>
          <div
                    className="loader"
          >
                    <img
                              alt="loading"
                              height={80}
                              src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                              width={80}
                    />
          </div>
          
</div>,
      ],
      "className": "pearon-users",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Pearson User Management ",
          "className": "heading",
        },
        "ref": null,
        "rendered": "Pearson User Management ",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="loader"
>
              <img
                            alt="loading"
                            height={80}
                            src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                            width={80}
              />
</div>,
            "",
          ],
          "className": "user-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="loading"
                height={80}
                src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                width={80}
/>,
              "className": "loader",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "loading",
                "height": 80,
                "src": "https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif",
                "width": 80,
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          "",
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="heading"
>
            Pearson User Management 
</h1>,
          <div
            className="user-container"
>
            <div
                        className="loader"
            >
                        <img
                                    alt="loading"
                                    height={80}
                                    src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                                    width={80}
                        />
            </div>
            
</div>,
        ],
        "className": "pearon-users",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Pearson User Management ",
            "className": "heading",
          },
          "ref": null,
          "rendered": "Pearson User Management ",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="loader"
>
                <img
                                alt="loading"
                                height={80}
                                src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                                width={80}
                />
</div>,
              "",
            ],
            "className": "user-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="loading"
                  height={80}
                  src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                  width={80}
/>,
                "className": "loader",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "loading",
                  "height": 80,
                  "src": "https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif",
                  "width": 80,
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "div",
            },
            "",
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
